Move case accessors into CommCareCase.objects - part 5

millerdev opened the pull request:
Continuation of https://github.com/dimagi/commcare-hq/pull/31051, https://github.com/dimagi/commcare-hq/pull/31060, https://github.com/dimagi/commcare-hq/pull/31065, and https://github.com/dimagi/commcare-hq/pull/31083

Case accessor methods are moving into `CommCareCase.objects` (which is `CommCareCaseManager`) or the relevant model manager that makes the most sense. `CaseAccessor(s|SQL)` will eventually be removed.

:blowfish: Review by commit.

## Safety Assurance

### Safety story

In addition to preserving all of the old tests and adapting the old methods that are being phased out to call the new method, new tests (copies of the old tests adapted to the new structure) are being added for the new model manager.

### Automated test coverage

Yes.

### QA Plan

No QA.

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

millerdev committed:
CommCareCaseIndexManager.get_related_indices

millerdev committed:
CommCareCaseManager.get_closed_and_deleted_ids

millerdev committed:
CommCareCaseManager.get_modified_case_ids

millerdev committed:
CommCareCaseIndexManager.get_extension_case_ids

millerdev committed:
CommCareCaseIndexManager.get_extension_chain

millerdev committed:
Lint

millerdev committed:
CommCareCaseManager.get_cases_by_external_id

millerdev committed:
self.form_accessor -> XFormInstance.objects

millerdev committed:
get_cases_by_external_id -> get_case_by_external_id

millerdev committed:
Improve test cleanup

millerdev committed:
get_case_by_domain_hq_user_id -> get_case_by_external_id

millerdev committed:
Use addClassCleanup rather than tearDownClass

It is more reliable: cleanups are called even when setUpClass fails

millerdev committed:
Check for null case

AmitPhulera commented on corehq/apps/case_importer/util.py:
```diff
@@ -146,7 +145,6 @@ def lookup_case(search_field, search_id, domain, case_type):
     error code (if there was an error in lookup).
```
> Attempt to find the case in CouchDB by the provided search_field and search_id.

I think this line does not hold true anymore, should we remove it?

millerdev replied:
Oh, right I should `s/in CouchDB//`. Thanks for flagging. I'll do that in a follow-up commit.

millerdev replied:
@AmitPhulera asked offline if there were tests for this `lookup_case` function. All tests that use this function mock the result, so tests were added in https://github.com/dimagi/commcare-hq/pull/31092/commits/1a269b81073f819e5e0050ab328a080ef9511870

AmitPhulera replied:
Thanks for adding the tests 😄

AmitPhulera commented on corehq/sql_db/tests/utils.py:
```diff
@@ -33,6 +39,22 @@ def __exit__(self, exc_type, exc_val, exc_tb):
             self.root_engine.dispose()
 
 
+def new_id_in_different_dbalias(partition_value):
```
Just for my understanding, what is the use-case that it would be used in?

millerdev replied:
It was added for this test

https://github.com/dimagi/commcare-hq/blob/b3c3124a10215be14b6bee6ebf3e3ca8b2fb540b/corehq/form_processor/tests/test_cases.py#L104-L120

millerdev committed:
Add tests for case_importer lookup_case(...)

These tests were also run against the pre-refactor version of the same function in commit d4b5c064a1f6b614248f5e317c134c09412195d5

To do that, test helpers must be copied from commit 06e86bb876c3bad710e1c7417673f40728d47ab9:

- corehq.form_processor.tests.test_cases._create_case
- corehq.sql_db.tests.utils.new_id_in_different_dbalias
